File: platform/wm/rhodes/phonebook/phonebook.cpp
/*------------------------------------------------------------------------
* (The MIT License)
*
* Copyright (c) 2008-2011 Rhomobile, Inc.
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*
* http://rhomobile.com
*------------------------------------------------------------------------*/

#include "stdafx.h"

#include "ext/rho/rhoruby.h"
#include "phonebook.h"

void* openPhonebook() {
#if defined (_WIN32_WCE) && !defined( OS_PLATFORM_MOTCE )
	CNativeAddressBook* phonebook = new CNativeAddressBook();

	if (!phonebook->openAB()) {
		delete phonebook;
		return NULL;
	}

	return phonebook;
#else
	return NULL;
#endif
}

void  closePhonebook(void* pb) {
#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
	if (pb) {
		CNativeAddressBook* phonebook = (CNativeAddressBook*)pb;
		delete phonebook;
	}
#endif
}

void _addRecordValue(const char* key, const char* value, void* hash) {
	addStrToHash(*((unsigned long*)hash), key, value);
}

#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
static VALUE _getRecord(CABRecord* record) {
	if (record) {
        CHoldRubyValue hash(rho_ruby_createHash());
		record->enumValues(_addRecordValue,&(hash.m_value));
		return hash;
	}
	return rho_ruby_get_NIL();
}
#endif

VALUE getallPhonebookRecords(void* pb) {
#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
	if (pb) {
		CNativeAddressBook* phonebook = (CNativeAddressBook*)pb;
        VALUE valGc = rho_ruby_disable_gc();
        CHoldRubyValue hash(rho_ruby_createHash());

		std::vector<CABRecord*> records;
		phonebook->getAllRecords(records);
		while(records.size() > 0) {
			CABRecord* record = records.back();
			const char* rid = record->getValue("id").c_str();
			if (rid) {
				LOGC(TRACE,"Phonebook") + "Adding record " + rid;
				addHashToHash(hash,rid,_getRecord(record));
			}
			delete record;
			records.pop_back();
		}

        rho_ruby_enable_gc(valGc);
		return hash;
	}
#endif
	return rho_ruby_get_NIL();
}

int getPhonebookRecordCount(void* pb) {
    return 0;
}

VALUE getPhonebookRecords(void* pb, int offset, int max_results, rho_param* select_param) {
	return rho_ruby_get_NIL();
}


VALUE getPhonebookRecord(void* pb, char* id) {
#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
	if (pb && id) {
		CNativeAddressBook* phonebook = (CNativeAddressBook*)pb;
		CABRecord* record = phonebook->getRecord(id);
		if (record) {
			VALUE hash = _getRecord(record);
			delete record;
			return hash;
		}
	}
#endif
	return rho_ruby_get_NIL();
}

//==================================================================================

VALUE getfirstPhonebookRecord(void* pb) {
	return rho_ruby_get_NIL();
}

VALUE getnextPhonebookRecord(void* pb) {
	return rho_ruby_get_NIL();}

//==================================================================================

void* createRecord(void* pb) {
#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
	return new CABOutlookRecord(NULL);
#else
	return NULL;
#endif
}

void* openPhonebookRecord(void* pb, char* id) {
#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
	if (pb && id) {
		CNativeAddressBook* phonebook = (CNativeAddressBook*)pb;
		return phonebook->getRecord(id);
	}
#endif
	return NULL;
}

int setRecordValue(void* record, char* prop, char* value) {
#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
	if (record) {
		CABRecord* r = (CABRecord*)record;
		r->setValue(prop,value);
	}
#endif
	return 1;
}

int addRecord(void* pb, void* record) {
#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
	if (pb && record) {
		CNativeAddressBook* phonebook = (CNativeAddressBook*)pb;
		phonebook->addRecord((CABRecord*)record);
		delete record;
	}
#endif
	return 1;
}

int saveRecord(void* pb, void* record) {
#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
	if (record) {
		CABRecord* r = (CABRecord*)record;
		LOGC(INFO,"Phonebook") + "About to save:";
		r->dump();
		r->save();
		delete  r;
	}
#endif
	return 1;
}

int deleteRecord(void* pb, void* record) {
#if defined (_WIN32_WCE)&& !defined( OS_PLATFORM_MOTCE )
	if (record) {
		CABRecord* r = (CABRecord*)record;
		r->remove();
		delete r;
	}
#endif
	return 1;
}

//==================================================================================

File: platform/android/Rhodes/src/com/rhomobile/rhodes/phonebook/Phonebook.java
/*------------------------------------------------------------------------
* (The MIT License)
* 
* Copyright (c) 2008-2011 Rhomobile, Inc.
* 
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
* 
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
* 
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
* 
* http://rhomobile.com
*------------------------------------------------------------------------*/

package com.rhomobile.rhodes.phonebook;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import android.os.Build;

import com.rhomobile.rhodes.Capabilities;
import com.rhomobile.rhodes.Logger;

//@RubyLevelClass(name="Phonebook")
public class Phonebook {

	private static final String TAG = "Phonebook";
	private static final boolean logging_enable = true;
	
	public static final String PB_ID = "id";
	public static final String PB_DISPLAY_NAME = "display_name";
	public static final String PB_FIRST_NAME = "first_name";
	public static final String PB_LAST_NAME = "last_name";
	public static final String PB_MOBILE_NUMBER = "mobile_number";
	public static final String PB_HOME_NUMBER = "home_number";
	public static final String PB_BUSINESS_NUMBER = "business_number";
	public static final String PB_EMAIL_ADDRESS = "email_address";
	public static final String PB_COMPANY_NAME = "company_name";
	
	private Map<String, Contact> contactList;
	private ContactAccessor accessor;
	private Iterator<Contact> iter = null;

	private boolean checkState() {
		if (!Capabilities.PIM_ENABLED)
			Logger.E(TAG, "Can not execute: PIM disabled");
		return Capabilities.PIM_ENABLED;
	}
	
	private ContactAccessor createAccessor() {
		String className;
		int sdkVersion = Integer.parseInt(Build.VERSION.SDK);
		if (sdkVersion < Build.VERSION_CODES.ECLAIR)
			className = "ContactAccessorOld";
		else
			className = "ContactAccessorNew";
		
		try {
			String pkgname = ContactAccessor.class.getPackage().getName();
			String fullName = pkgname + "." + className;
			Class<? extends ContactAccessor> klass =
				Class.forName(fullName).asSubclass(ContactAccessor.class);
			return klass.newInstance();
		}
		catch (Exception e) {
			throw new IllegalStateException(e);
		}
	}
	
	public Phonebook() {
		try {
			if (!checkState())
				return;
			
			accessor = createAccessor();
			contactList = new HashMap<String, Contact>();
		}
		catch (Exception e) {
			Logger.E(TAG, e);
		}
	}
	
	public int queryContactCount(int offset, int limit) {
		int res = 0;
		if (checkState()) {
			try {
				res = accessor.getCount(offset, limit);
			} catch (Exception e) {
				Logger.E(TAG, e);
			}
		}
		return res;
	}
	
	public void queryAllContacts() {
		String[] allColumns = new String[] {
			PB_ID,
			PB_DISPLAY_NAME,
			PB_FIRST_NAME,
			PB_LAST_NAME,
			PB_MOBILE_NUMBER,
			PB_HOME_NUMBER,
			PB_BUSINESS_NUMBER,
			PB_EMAIL_ADDRESS,
			PB_COMPANY_NAME };
		queryContacts(0, -1, Arrays.asList(allColumns));
	}

	public void queryContacts(int offset, int max_results, List<String> select) {
		if (!checkState())
			return;
		try {
			Logger.I(TAG, "queryContacts()");
			contactList = accessor.getContacts(offset, max_results, select);
		}
		catch (Exception e) {
			Logger.E(TAG, e);
		}
	}
	
	public void close() {
		try {
			if (!checkState())
				return;
			
			this.contactList.clear();
		}
		catch (Exception e) {
			Logger.E(TAG, e);
		}
	}

	public void moveToBegin() {
		try {
			if (!checkState())
				return;
			iter = contactList.values().iterator();
		}
		catch (Exception e) {
			Logger.E(TAG, e);
		}
	}
	
	public boolean hasNext() {
		try {
			if (!checkState())
				return false;

			return iter.hasNext();
		}
		catch (Exception e) {
			Logger.E(TAG, e);
			return false;
		}
	}
	
	public Object next() {
		try {
			if (!checkState())
				return null;
			
			return iter.next();
		}
		catch (Exception e) {
			Logger.E(TAG, e);
			return null;
		}
	}
	
	public Contact getFirstRecord() {
		try {
			if (!checkState())
				return null;
			
			moveToBegin();
			if (!iter.hasNext())
				return null;
			return iter.next();
		}
		catch (Exception e) {
			Logger.E(TAG, e);
			return null;
		}
	}
	
	public Contact getNextRecord() {
		try {
			if (!checkState())
				return null;
			
			return iter.next();
		}
		catch (Exception e) {
			Logger.E(TAG, e);
			return null;
		}
	}
	
	public Contact getRecord(String idd) {
		if (logging_enable) Logger.I(TAG, "Phonebook.getRecord("+idd+")");
		try {
			if (!checkState())
				return null;

			Contact cc = accessor.getContact(idd);
			if (cc == null)
				Logger.W(TAG, "Phonebook.getRecord() contact record not found.");
				
			return cc;
		}
		catch (Exception e) {
			Logger.E(TAG, e);
			return null;
		}
	}

	public void removeContact(Contact contact) {
		try {
			if (!checkState())
				return;
			
			accessor.remove(contact);
			contactList.remove(contact.getField(Phonebook.PB_ID));
		}
		catch (Exception e) {
			Logger.E(TAG, e);
		}
	}

	public void saveContact(Contact contact) {
		try {
			if (!checkState())
				return;
			
			accessor.save(contact);
			contactList.put(contact.getField(Phonebook.PB_ID), contact);
		}
		catch (Exception e) {
			Logger.E(TAG, e);
		}
	}
}
